File: zkd-android/src/lib/time.ts
/**
 * Everything is derived from the real clock so the disruption is always
 * happening right now. All of it takes `now` explicitly — nothing reads the
 * clock at module scope, because that would differ between the server render
 * and the client render and blow up hydration.
 */

export const mins = (d: Date, m: number) => new Date(d.getTime() + m * 60_000);
export const days = (d: Date, n: number) => {
  const x = new Date(d);
  x.setDate(x.getDate() + n);
  return x;
};
export const hhmm = (d: Date) =>
  `${String(d.getHours()).padStart(2, '0')}:${String(d.getMinutes()).padStart(2, '0')}`;
export const hhmmss = (d: Date) => `${hhmm(d)}:${String(d.getSeconds()).padStart(2, '0')}`;
export const ddMon = (d: Date) =>
  `${d.getDate()} ${d.toLocaleString('en-GB', { month: 'short' })}`;

const midnight = (d: Date) => {
  const x = new Date(d);
  x.setHours(0, 0, 0, 0);
  return x;
};

export function dayLabel(d: Date, now: Date) {
  const diff = Math.round((midnight(d).getTime() - midnight(now).getTime()) / 86_400_000);
  if (diff === 0) return `Today, ${ddMon(d)}`;
  if (diff === 1) return `Tomorrow, ${ddMon(d)}`;
  if (diff === -1) return 'Yesterday';
  return ddMon(d) + (d.getFullYear() !== now.getFullYear() ? ` ${d.getFullYear()}` : '');
}

export function agoLabel(d: Date, now: Date) {
  const s = Math.max(0, Math.round((now.getTime() - d.getTime()) / 1000));
  if (s < 60) return `${s} seconds ago`;
  const m = Math.round(s / 60);
  if (m < 60) return `${m} ${m === 1 ? 'minute' : 'minutes'} ago`;
  const h = Math.round(m / 60);
  if (h < 24) return `${h} ${h === 1 ? 'hour' : 'hours'} ago`;
  const dd = Math.round(h / 24);
  if (dd < 14) return `${dd} ${dd === 1 ? 'day' : 'days'} ago`;
  if (dd < 70) return `${Math.round(dd / 7)} weeks ago`;
  const mo = Math.round(dd / 30);
  if (mo < 18) return `${mo} months ago`;
  const y = (dd / 365).toFixed(1).replace('.0', '');
  return `${y} ${y === '1' ? 'year' : 'years'} ago`;
}

export const durLabel = (m: number) =>
  `${Math.floor(m / 60)}h ${String(m % 60).padStart(2, '0')}m`;

/** seconds, printed the way an engineer would read a latency budget */
export const secs = (n: number) =>
  `${(n < 1 ? n.toFixed(2) : n.toFixed(1)).replace(/\.0$/, '')}s`;

export const money = (n: number) => (n ? `₹${n.toLocaleString('en-IN')}` : '₹0');

File: zkd-flutter/lib/util/time.dart
/// Display formatters, ported from zkd-android/src/lib/time.ts.
///
/// Hand-rolled rather than delegated to intl's locale data: these need to match
/// the web app's output exactly, and the money format is Indian digit grouping
/// (lakh/crore), which is not the same as splitting on every three digits.
const List<String> _months = [
  'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
  'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec',
];

String _p2(int n) => n.toString().padLeft(2, '0');

String hhmm(DateTime d) => '${_p2(d.hour)}:${_p2(d.minute)}';

String ddMon(DateTime d) => '${d.day} ${_months[d.month - 1]}';

DateTime _midnight(DateTime d) => DateTime(d.year, d.month, d.day);

/// "Today, 19 Aug" / "Tomorrow, 19 Aug" / "Yesterday" / "19 Aug 2027"
String dayLabel(DateTime d, DateTime now) {
  final diff = _midnight(d).difference(_midnight(now)).inDays;
  if (diff == 0) return 'Today, ${ddMon(d)}';
  if (diff == 1) return 'Tomorrow, ${ddMon(d)}';
  if (diff == -1) return 'Yesterday';
  return ddMon(d) + (d.year != now.year ? ' ${d.year}' : '');
}

/// "42 seconds ago", "3 minutes ago", "2 hours ago", "5 days ago", "1.5 years ago"
String agoLabel(DateTime d, DateTime now) {
  final s = now.difference(d).inSeconds;
  final sec = s < 0 ? 0 : s;
  if (sec < 60) return '$sec seconds ago';
  final m = (sec / 60).round();
  if (m < 60) return '$m ${m == 1 ? 'minute' : 'minutes'} ago';
  final h = (m / 60).round();
  if (h < 24) return '$h ${h == 1 ? 'hour' : 'hours'} ago';
  final dd = (h / 24).round();
  if (dd < 14) return '$dd ${dd == 1 ? 'day' : 'days'} ago';
  if (dd < 70) return '${(dd / 7).round()} weeks ago';
  final mo = (dd / 30).round();
  if (mo < 18) return '$mo months ago';
  final y = (dd / 365).toStringAsFixed(1).replaceAll(RegExp(r'\.0$'), '');
  return '$y ${y == '1' ? 'year' : 'years'} ago';
}

/// "2h 05m"
String durLabel(int minutes) => '${minutes ~/ 60}h ${_p2(minutes % 60)}m';

/// Seconds the way an engineer reads a latency budget: "0.40s", "3.4s", "42s".
String secs(num n) {
  final t = n < 1 ? n.toStringAsFixed(2) : n.toStringAsFixed(1);
  return '${t.replaceAll(RegExp(r'\.0$'), '')}s';
}

/// Indian digit grouping: 1234567 -> "12,34,567". Zero is "₹0", never "".
String money(num n) {
  if (n == 0) return '₹0';
  final neg = n < 0;
  final digits = n.abs().round().toString();
  String grouped;
  if (digits.length <= 3) {
    grouped = digits;
  } else {
    final last3 = digits.substring(digits.length - 3);
    var rest = digits.substring(0, digits.length - 3);
    final parts = <String>[];
    while (rest.length > 2) {
      parts.insert(0, rest.substring(rest.length - 2));
      rest = rest.substring(0, rest.length - 2);
    }
    if (rest.isNotEmpty) parts.insert(0, rest);
    grouped = '${parts.join(',')},$last3';
  }
  return '₹${neg ? '-' : ''}$grouped';
}
